File: client-dotnet/src/EvoMaster.Instrumentation/Coverage/MethodReplacement/ObjectClassReplacement.cs
using System;
using EvoMaster.Client.Util.Extensions;
using EvoMaster.Instrumentation.Heuristic;
using EvoMaster.Instrumentation.StaticState;
using EvoMaster.Instrumentation_Shared;

namespace EvoMaster.Instrumentation.Coverage.MethodReplacement{
    public class ObjectClassReplacement{
        public static bool EqualsObject(object left, object right, string idTemplate){
            ObjectExtensions.RequireNonNull<object>(left);
            
            bool result = left.Equals(right);

            if (idTemplate == null){
                return result;
            }

            Truthness t;

            if (result){
                t = new Truthness(1d, DistanceHelper.H_NOT_NULL);
            } else{
                if (right == null){
                    t = new Truthness(DistanceHelper.H_REACHED_BUT_NULL, 1d);
                } else{
                    double baseS = DistanceHelper.H_NOT_NULL;
                    double distance = DistanceHelper.GetDistance(left, right);
                    double h = DistanceHelper.HeuristicFromScaledDistanceWithBase(baseS, distance);
                    t = new Truthness(h, 1d);
                }
            }

            ExecutionTracer.ExecutedReplacedMethod(idTemplate, ReplacementType.BOOLEAN, t);

            return result;
        }
    }
}

File: client-java/instrumentation/src/main/java/org/evomaster/client/java/instrumentation/coverage/methodreplacement/classes/ObjectClassReplacement.java
package org.evomaster.client.java.instrumentation.coverage.methodreplacement.classes;

import org.evomaster.client.java.instrumentation.coverage.methodreplacement.MethodReplacementClass;
import org.evomaster.client.java.instrumentation.coverage.methodreplacement.Replacement;
import org.evomaster.client.java.instrumentation.shared.ReplacementCategory;
import org.evomaster.client.java.instrumentation.shared.ReplacementType;

public class ObjectClassReplacement implements MethodReplacementClass {


    @Override
    public Class<?> getTargetClass() {
        return Object.class;
    }

    @Replacement(type = ReplacementType.BOOLEAN, category = ReplacementCategory.EXT_0)
    public static boolean equals(Object left, Object right, String idTemplate) {

        if(left == null) {
           left.equals(right); //throw NPE
        }

        return ObjectsClassReplacement.equals(left, right, idTemplate);
    }

}
